File: remote/java/src/main/java/org/pxscene/rt/remote/RTRemoteConnectionManager.java
package org.pxscene.rt.remote;

import java.net.URI;
import java.util.HashMap;
import java.util.Map;
import org.apache.log4j.Logger;
import org.pxscene.rt.RTException;
import org.pxscene.rt.RTObject;

/**
 * The remote connection manager class.
 */
public class RTRemoteConnectionManager {

  /**
   * the connections map object
   */
  private static final Map<String, RTRemoteClientConnection> connections = new HashMap<>();

  /**
   * the logger instance
   */
  private static final Logger logger = Logger.getLogger(RTRemoteConnectionManager.class);

  /**
   * get remote object by uri
   *
   * @param uri the object uri
   * @return the remote
   * @throws RTException if any other error occurred during operation
   */
  public static RTObject getObjectProxy(URI uri) throws RTException {
    RTRemoteClientConnection connection;

    String connectionSpec = uri.getScheme() + "//" + uri.getHost() + ":" + uri.getPort();
    if (connections.containsKey(connectionSpec)) {
      connection = connections.get(connectionSpec);
    } else {
      connection = RTRemoteConnectionManager.createConnectionFromSpec(uri);
      connections.put(connectionSpec, connection);
    }

    return connection.getProxyObject(uri.getPath().substring(1));
  }

  /**
   * create connection from uri
   *
   * @param uri the dest uri
   * @return the RTRemoteClientConnection
   * @throws RTException if any other error occurred during operation
   */
  private static RTRemoteClientConnection createConnectionFromSpec(URI uri) throws RTException {
    RTRemoteClientConnection connection;
    logger.info("start connection " + uri.toString());
    switch (uri.getScheme()) {
      case "tcp":
        connection = RTRemoteClientConnection
            .createTCPClientConnection(uri.getHost(), uri.getPort());
        break;
      default:
        throw new RTException("unsupported scheme:" + uri.getScheme());
    }
    return connection;
  }
}

File: remote/js/lib/RTRemoteConnectionManager.js
/*
 pxCore Copyright 2005-2018 John Robinson

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

/**
 * The remote connection manager
 *
 * @author      TCSCODER
 * @version     1.0
 */
const { URL } = require('url');
const logger = require('./common/logger');
const RTRemoteClientConnection = require('./RTRemoteClientConnection');
const RTException = require('./RTException');

/**
 * the connections map
 * @type {object} the connections map
 */
const connections = {};

/**
 * get rt remote object by uri
 * @param {string} uri the object uri
 * @return {Promise<RTRemoteObject>} the promise with rt remote object
 */
function getObjectProxy(uri) {
  const url = new URL(uri);
  const connectionSpec = `${url.protocol}//${url.hostname}:${url.port}`;
  const getRemoteObject = (conn, pathname) => conn.getProxyObject(pathname.substr(1, pathname.length));
  if (connections[connectionSpec]) {
    return Promise.resolve(getRemoteObject(connections[connectionSpec], url.pathname));
  }
  return createConnectionFromSpec(url).then((connection) => {
    connections[connectionSpec] = connection;
    return Promise.resolve(getRemoteObject(connection, url.pathname));
  });
}

/**
 * create connection from url
 * @param {URL} url the dest url the url object
 * @return {Promise<RTRemoteClientConnection>} the promise with connection
 */
function createConnectionFromSpec(url) {
  let connection = null;
  const schema = url.protocol.substr(0, url.protocol.length - 1);
  logger.info(`start connection ${url}`);
  switch (schema) {
    case 'tcp':
      connection = RTRemoteClientConnection.createTCPClientConnection(url.hostname, url.port);
      break;
    default:
      throw new RTException(`unsupported scheme : ${url.protocol}`);
  }
  return connection;
}

module.exports = {
  getObjectProxy,
};
